### PR TITLE
fix(pipelines-widget): clean up observable usage

### PR DESCRIPTION
**Pull request by jiekang**
This small PR cleans up some observable usage in the pipelines widget found when trying to diagnose https://github.com/openshiftio/openshift.io/issues/2744

It's purely a code cleanup, no bug fix or functionality improvement.

### CODE REVIEW

**Comment by andrewazores** [comment]
Looks fine. What do you think of removing the `buildConfigsCount` component property and just using `.length` in the template wherever that property was used instead?

**Comment by jiekang** [comment]
@andrewazores Just noticed with these changes I get

`Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: '[object Object]'. Current value: 'true'.`

on

`<div class="f8-blank-slate-card" *ngIf="!(buildConfigsCount | async)">`]

And remains after change to use `(buildConfigs | async).length`

**Comment by jiekang** [comment]
I've investigated the change detection issue and resolved it. This can be reviewed again; sorry about that.

File: src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.html [hunk 1]
```diff
@@ -6,7 +6,7 @@ <h2 class="card-pf-title">
       </h2>
     </div>
     <div class="card-pf-body f8-card-body">
-      <div class="f8-blank-slate-card" *ngIf="!(buildConfigsCount | async)">
+      <div class="f8-blank-slate-card" *ngIf="1 > buildConfigsCount; else appsAvailable">
```
Comment (andrewazores) [inline src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.html]: `1 > buildConfigsCount`? This is a bit awkward to read... why not `buildConfigsCount < 1` or `buildConfigsCount === 0` even?

File: src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.html [hunk 2]
```diff
@@ -104,11 +106,11 @@ <h4 class="f8-card-applications-list-step">Run</h4>
       </div>
       <h2 class="card-pf-title">
         <a id="spacehome-pipelines-title" [routerLink]="[contextPath | async, 'create', 'pipelines']">Pipelines</a>
-        <span id="spacehome-pipelines-badge" class="badge f8-card-badge">{{buildConfigsCount | async}}</span>
+        <span id="spacehome-pipelines-badge" class="badge f8-card-badge">{{buildConfigsCount}}</span>
       </h2>
     </div>
     <div class="card-pf-body f8-card-body">
-      <div class="f8-blank-slate-card" *ngIf="!(buildConfigsCount | async)">
+      <div class="f8-blank-slate-card" *ngIf="1 > buildConfigsCount; else pipeAvailable">
```
Comment (andrewazores) [inline src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.html]: Same comment here re. `1 > buildConfigsCount`

File: src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.ts [hunk 3]
```diff
@@ -40,10 +39,13 @@ export class PipelinesWidgetComponent implements OnInit {
...
-      .publish();
-    this.buildConfigsCount = this.buildConfigs.map(buildConfigs => buildConfigs.length);
-    this.buildConfigs.connect();
+    this.buildConfigs = this.pipelinesService.current.share();
+    // buildConfigsCount triggers changes in the DOM; force Angular Change Detection
+    // via setTimeout encapsulation
+    this.buildConfigs
+      .map(buildConfigs => buildConfigs.length)
+      .share()
+      .subscribe(length => setTimeout(() => this.buildConfigsCount = length));
```
Comment (andrewazores) [inline src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.ts]: Did we figure out how the ConnectableObservable was working around the change detection? This makes sense but I'm not clear on how/why it was working before...

Comment (jiekang) [reply src/app/dashboard-widgets/pipelines-widget/pipelines-widget.component.ts]: I can only surmise it's purely a coincidence with the timing. Removing the Connectable publish/connect, I can refresh the page multiple times and, in the majority see a CD error and a minority, less than 1/10, not see it.

**Pull request merged**